### ISSUE TITLE
Is CRTP synonymous to static polymorphism in C++

### ISSUE THREAD

**Issue by haptork** [lorenzhs/cpp-static-polymorphism-proposal#1]
As of my current understanding, polymorphism is when a function/class has multiple implementations to select from. Static polymorphism is when the choice is made at compile time. According to these definitions I can think of following type of static polymorphisms in C++:

1. function overloading
2. template specialisation for specific types
3. template specialisation with SFINAE / enable if
4. template class method specialisation
5. CRTP

I think there can be more ways or may be some of the ways can be combined together. I think of 2 and 3 points whenever static polymorphism is mentioned for C++. 

The CRTP idiom is also termed theoretically as f-bounded polymorphism.

Does this proposal include all sort of template static polymorphism or just CRTP?

**Comment by lorenzhs** [comment]
Right, static polymorphism is more than just CRTP. I'm only talking about SP with subtyping here. The items of your list serve different purposes:

1. *function overloading* - useful to distinguish behaviour of a function based on arguments (types or count)
2. *template specialisation* - useful for modifying behaviour based on a template parameter
3. *SFINAE / `enable_if`* - similar to the above
4. *template class method specialisation* - depending on how you implement it that's pretty much 1 or 2
5. *CRTP* - useful e.g. if you want to have common code in a base class that calls a method of the derived class, resolved at compile time.

So CRTP serves a very different purpose from the others, and the suggestion wouldn't be applicable to the other types.

I'll try to clarify the terminology. Thanks for bringing this to my attention!

**Comment by haptork** [comment]
Yes, CRTP seems to be irreplacable by other techniques at times. However, a lot of cases where goal is to seek static polymorphism can be modeled in any of these ways. Look at the different solutions given to the following SO questions:
- http://stackoverflow.com/q/36629722/1583711
- http://stackoverflow.com/a/36276703/1583711

These cases are the original source of my confusion.

**Comment by lorenzhs** [comment]
Hm. The use case that triggered this was that I have many strategies that share some common support code, and must implement certain methods (but may override default implementations of others). This is quite easy with CRTP, but could look nicer. A trimmed-down example could look like this:

```cpp
template <typename Strategy>
struct FooBase {
    bool condition(int x) { return x % 2 == 0; }
    int foo(int x) {
        if (static_cast<Strategy*>(this)->condition(x)) { return x; }
        else { return static_cast<Strategy*>(this)->blah(x); }
    }
};

struct Raboof : public FooBase<Raboof> {
    int blah(int x) { return 2 * x; }
};

struct Bar : public FooBase<Bar> {
    bool condition(int x) { return x < 0; }
    int blah(int x) { return x * x; }
};
```
Then I can use `Raboof` and `Bar` without any visible templatey things, with everything well hidden behind the curtain.

This works very well with CRTP, but requiring that seems a bit arcane. It would be much nicer if I could explicitly mark `blah` as deferred.

But while typing this I noticed that my suggestion doesn't work well with default implementations. Basically, a keyword (e.g. `deferred`) would be needed that marks a method as *"use the most specialised implementation of this"*, which CRTP accomplishes with the cast.

I think I need to give this some more thought. But right now, it seems that this would only make sense with a new keyword (like `deferred`, used the same way as `virtual`). That could possibly be quite confusing, though. Maybe CRTP isn't so bad after all?

**Comment by haptork** [comment]
I like the initiative and it will certainly encourage the use of static polymorphism over dynamic. However, I bring this up just to throw some light on cases where CRTP can be substituted with some other static binding technique (for better or worse).

**Comment by lorenzhs** [comment]
Thanks for that, you helped me find a gap in my idea :) It's clear that CRTP isn't the only mechanism to achieve static polymorphism in C++, but the appropriate mechanism varies a lot between use cases. It's good to have a more complete list. I'll update this when I have an idea.
